File: system/sensord/sensors/mmc5603nj_magn.cc
#include "system/sensord/sensors/mmc5603nj_magn.h"

#include <algorithm>
#include <cassert>
#include <vector>

#include "common/swaglog.h"
#include "common/timing.h"
#include "common/util.h"

MMC5603NJ_Magn::MMC5603NJ_Magn(I2CBus *bus) : I2CSensor(bus) {}

int MMC5603NJ_Magn::init() {
  int ret = verify_chip_id(MMC5603NJ_I2C_REG_ID, {MMC5603NJ_CHIP_ID});
  if (ret == -1) return -1;

  // Set ODR to 0
  ret = set_register(MMC5603NJ_I2C_REG_ODR, 0);
  if (ret < 0) {
    goto fail;
  }

  // Set BW to 0b01 for 1-150 Hz operation
  ret = set_register(MMC5603NJ_I2C_REG_INTERNAL_1, 0b01);
  if (ret < 0) {
    goto fail;
  }

fail:
  return ret;
}

int MMC5603NJ_Magn::shutdown() {
  int ret = 0;

  // disable auto reset of measurements
  uint8_t value = 0;
  ret = read_register(MMC5603NJ_I2C_REG_INTERNAL_0, &value, 1);
  if (ret < 0) {
    goto fail;
  }

  value &= ~(MMC5603NJ_CMM_FREQ_EN | MMC5603NJ_AUTO_SR_EN);
  ret = set_register(MMC5603NJ_I2C_REG_INTERNAL_0, value);
  if (ret < 0) {
    goto fail;
  }

  // set ODR to 0 to leave continuous mode
  ret = set_register(MMC5603NJ_I2C_REG_ODR, 0);
  if (ret < 0) {
    goto fail;
  }
  return ret;

fail:
  LOGE("Could not disable mmc5603nj auto set reset");
  return ret;
}

void MMC5603NJ_Magn::start_measurement() {
  set_register(MMC5603NJ_I2C_REG_INTERNAL_0, 0b01);
  util::sleep_for(5);
}

std::vector<float> MMC5603NJ_Magn::read_measurement() {
  int len;
  uint8_t buffer[9];
  len = read_register(MMC5603NJ_I2C_REG_XOUT0, buffer, sizeof(buffer));
  assert(len == sizeof(buffer));
  float scale = 1.0 / 16384.0;
  float x = (read_20_bit(buffer[6], buffer[1], buffer[0]) * scale) - 32.0;
  float y = (read_20_bit(buffer[7], buffer[3], buffer[2]) * scale) - 32.0;
  float z = (read_20_bit(buffer[8], buffer[5], buffer[4]) * scale) - 32.0;
  std::vector<float> xyz = {x, y, z};
  return xyz;
}

bool MMC5603NJ_Magn::get_event(MessageBuilder &msg, uint64_t ts) {
  uint64_t start_time = nanos_since_boot();
  // SET - RESET cycle
  set_register(MMC5603NJ_I2C_REG_INTERNAL_0, MMC5603NJ_SET);
  util::sleep_for(5);
  MMC5603NJ_Magn::start_measurement();
  std::vector<float> xyz = MMC5603NJ_Magn::read_measurement();

  set_register(MMC5603NJ_I2C_REG_INTERNAL_0, MMC5603NJ_RESET);
  util::sleep_for(5);
  MMC5603NJ_Magn::start_measurement();
  std::vector<float> reset_xyz = MMC5603NJ_Magn::read_measurement();

  auto event = msg.initEvent().initMagnetometer();
  event.setSource(cereal::SensorEventData::SensorSource::MMC5603NJ);
  event.setVersion(1);
  event.setSensor(SENSOR_MAGNETOMETER_UNCALIBRATED);
  event.setType(SENSOR_TYPE_MAGNETIC_FIELD_UNCALIBRATED);
  event.setTimestamp(start_time);

  float vals[] = {xyz[0], xyz[1], xyz[2], reset_xyz[0], reset_xyz[1], reset_xyz[2]};
  bool valid = true;
  if (std::any_of(std::begin(vals), std::end(vals), [](float val) { return val == -32.0; })) {
    valid = false;
  }
  auto svec = event.initMagneticUncalibrated();
  svec.setV(vals);
  svec.setStatus(valid);
  return true;
}

File: system/sensord/sensors/mmc5603nj_magn.py
import time

from cereal import log
from openpilot.system.sensord.sensors.i2c_sensor import Sensor

# https://www.mouser.com/datasheet/2/821/Memsic_09102019_Datasheet_Rev.B-1635324.pdf

# Register addresses
REG_ODR = 0x1A
REG_INTERNAL_0 = 0x1B
REG_INTERNAL_1 = 0x1C

# Control register settings
CMM_FREQ_EN = (1 << 7)
AUTO_SR_EN  = (1 << 5)
SET         = (1 << 3)
RESET       = (1 << 4)

class MMC5603NJ_Magn(Sensor):
  @property
  def device_address(self) -> int:
    return 0x30

  def init(self):
    self.verify_chip_id(0x39, [0x10, ])
    self.writes((
      (REG_ODR, 0),

      # Set BW to 0b01 for 1-150 Hz operation
      (REG_INTERNAL_1, 0b01),
    ))

  def _read_data(self, cycle) -> list[float]:
    # start measurement
    self.write(REG_INTERNAL_0, cycle)
    self.wait()

    # read out XYZ
    scale = 1.0 / 16384.0
    b = self.read(0x00, 9)
    return [
      (self.parse_20bit(b[6], b[1], b[0]) * scale) - 32.0,
      (self.parse_20bit(b[7], b[3], b[2]) * scale) - 32.0,
      (self.parse_20bit(b[8], b[5], b[4]) * scale) - 32.0,
    ]

  def get_event(self, ts: int | None = None) -> log.SensorEventData:
    ts = time.monotonic_ns()

    # SET - RESET cycle
    xyz = self._read_data(SET)
    reset_xyz = self._read_data(RESET)
    vals = [*xyz, *reset_xyz]

    event = log.SensorEventData.new_message()
    event.timestamp = ts
    event.version = 1
    event.sensor = 3 # SENSOR_MAGNETOMETER_UNCALIBRATED
    event.type = 14  # SENSOR_TYPE_MAGNETIC_FIELD_UNCALIBRATED
    event.source = log.SensorEventData.SensorSource.mmc5603nj

    m = event.init('magneticUncalibrated')
    m.v = vals
    m.status = int(all(int(v) != -32 for v in vals))

    return event

  def shutdown(self) -> None:
    v = self.read(REG_INTERNAL_0, 1)[0]
    self.writes((
      # disable auto-reset of measurements
      (REG_INTERNAL_0, (v & (~(CMM_FREQ_EN | AUTO_SR_EN)))),

      # disable continuous mode
      (REG_ODR, 0),
    ))
